### ISSUE TITLE
Foreman SIGTERM won't shut down workers?

### ISSUE THREAD

**Issue by masterkain** [mperham/sidekiq#85]
Hello,
I'm using Foreman with this simple Procfile in development, using Rails 3.2.2 and sidekiq 0.10.0 (git):

```
web: bundle exec thin start -p $PORT --stats /thin
workers: bundle exec sidekiq -c 2
```

With a clean Redis database I issue `foreman start` and the defined commands gets executed, I then check the web-ui and see two workers.

I then interrupt foreman with a `ctrl+c`.
Foreman says he got a `SIGINT` and that a `SIGTERM` and `SIGKILL` is being sent to other processes (exception not always thrown):

```
...
05:57:18 workers.1  | 2012-03-19T04:57:18Z 62825 TID-oxrlnco24 INFO: Booting sidekiq 0.10.0 with Redis at localhost:6379
05:57:18 workers.1  | 2012-03-19T04:57:18Z 62825 TID-oxrlnco24 INFO: Running in ruby 1.9.3p125 (2012-02-16 revision 34643) [x86_64-darwin11.3.0]
^CSIGINT received
05:57:45 system     | sending SIGTERM to all processes
05:57:45 system     | sending SIGTERM to pid 62824
05:57:45 web.1      | >> Stopping ...
05:57:45 web.1      | >> Stopping ...
05:57:45 system     | sending SIGTERM to pid 62825
05:57:45 workers.1  | /Users/kain/.rvm/gems/ruby-1.9.3-p125/bundler/gems/sidekiq-fe49a425c806/lib/sidekiq/cli.rb:9:in `block in <top (required)>': Interrupt (Interrupt)
05:57:45 workers.1  |   from /Users/kain/.rvm/rubies/ruby-1.9.3-p125/lib/ruby/1.9.1/timeout.rb:59:in `call'
05:57:45 workers.1  |   from /Users/kain/.rvm/rubies/ruby-1.9.3-p125/lib/ruby/1.9.1/timeout.rb:59:in `timeout'
05:57:45 workers.1  |   from /Users/kain/.rvm/gems/ruby-1.9.3-p125/gems/celluloid-0.9.0/lib/celluloid.rb:63:in `shutdown'
05:57:45 workers.1  |   from /Users/kain/.rvm/gems/ruby-1.9.3-p125/gems/celluloid-0.9.0/lib/celluloid.rb:82:in `block in <module:Celluloid>'
05:57:46 web.1      | process terminated
05:57:50 system     | sending SIGKILL to all processes
05:57:50 system     | sending SIGKILL to pid 62825
05:57:50 workers.1  | process terminated
```

Now, if I run `foreman start` again and check to the web-ui interface, I can see the number of workers, both in the counter and the list, is now four, and it increases at every startup.

Today I found sidekiq frozen consuming 100% cpu, apparently classified as a zombie process since the parent was already gone, not sure if it can be related to this but worth mentioning.

**Comment by jfelchner** [comment]
Same problem here.  This should probably be reopened.  I eventually end up with about 5 sidekiq processes running due to it not shutting down.  I have to send `kill -9` rather than the regular `kill` in order for it to shut down.

**Comment by mperham** [comment]
Try latest version.

**Comment by jfelchner** [comment]
Will do and I'll report back. Thanks!

**Comment by jfelchner** [comment]
Haven't been having any issues but the problem was intermittent so I'll ping here if it happens again.

**Comment by localshred** [comment]
Old thread, but FWIW I was having this same issue on the latest Sidekiq (3.4.2) using sidetiq 0.6.3. Foreman wouldn't send SIGTERM to sidekiq when using sidtiq. I've [submitted an issue over there](https://github.com/tobiassvn/sidetiq/issues/137) to see if I can help clean this up.

**Comment by jfelchner** [comment]
Still having this issue on `3.4.2`  It just happened today.

**Comment by mperham** [comment]
There's an open foreman issue. It's not a Sidekiq issue.

**Comment by jfelchner** [comment]
@mperham awesome!  Thanks!

**Comment by lorrin** [comment]
After spending some time poking at this, I believe that Foreman and Sidekiq _both_ have bugs. To cause both bugs to surface, have foreman invoke a shell script that starts Sidekiq. When you Ctrl-C foreman, the following sequence of events happens:
1. the SIGINT passes through foreman to its children, and Sidekiq begins its graceful shutdown. Log output:
```
16:40:23 sidekiq.1  | 2023-08-11T23:40:23.776Z pid=25869 tid=h1p INFO: Shutting down
16:40:23 sidekiq.1  | 2023-08-11T23:40:23.776Z pid=25869 tid=h1p INFO: Terminating quiet threads for default capsule
16:40:23 sidekiq.1  | 2023-08-11T23:40:23.787Z pid=25869 tid=1hip INFO: Scheduler exiting...
16:40:23 sidekiq.1  | 2023-08-11T23:40:23.788Z pid=25869 tid=1hdl INFO: Leader stepping down
```
2. Foreman immediately begins its graceful termination, which involves sending a SIGTERM to all the children
1. The shell script exits upon receiving the SIGTERM
1. foreman sees that the shell script has exited and logs that it was `terminated by SIGTERM`
1. foreman is completely unaware of Sidekiq. It exits.
1. Sidekiq lives on forever and never terminates. In `ps`, it shows up as `sidekiq 7.1.0 app_name [0 of 5 busy] stopping`. Its state is `R` (runnable) in ps it spins up to 100% CPU usage on one core. So we're not talking about a zombie process description that's harmlessly hanging around because it was never properly reaped.

Obviously, foreman could do better and that's why there are so many forks of it that use process groups and don't have this problem. But regardless of foreman's shortcomings, I don't know what could explain the Sidekiq process remaining running (and using 100% CPU) other than a Sidekiq bug. My hunch is that when foreman exits, Sidekiq's stdout/stderr get closed, and this causes it to get stuck trying to log. I did not do a reproduction in which I directly ran sidekiq and somehow closed its stdout/stderr without involving foreman. But I did change my shell script to run `sidekiq > sidekiq.log 2>&1`. In that case the observed behavior was that foreman still prematurely exited, but Sidekiq eventually completed its graceful termination and exited. That's some anecdotal evidence that stdout/stderr is the issue. 

This behavior was observed on multiple developer machines. All were M1/M2 Macs.

For anyone stumbling on this and looking for a workaround, one option is to change your shell script to `exec sidekiq`. That will cause sidekiq to replace bash as foreman's child. That sidesteps foreman's bug in managing grandchildren. Another is to use [hivemind](https://github.com/openhive-network/hivemind) or one of the other drop-in replacements for foreman.
